### PR TITLE
add a little bit code and results

### PR DESCRIPTION
**Pull request by amurzina**

### CODE REVIEW

File: content/posts/detecting-web-attacks-rnn.md [hunk 1]
```diff
@@ -27,21 +27,33 @@ used to back in a day. But does it?
...
+- Path Traversal  
+- OS Commanding
+- Object injection
+- etc.
+
+
 But first let's ask ourselves:
 
-### What would a human do?
+### What would a human do when he sees a single request?
```
Comment (montekki) [inline content/posts/detecting-web-attacks-rnn.md]: better delete this, also not sure if proper grammar

File: content/posts/detecting-web-attacks-rnn.md [hunk 2]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
 is given an anomalous sample it is likely to re-create it with a high degree
 of error.
 
+So, basically, autoencoder model consists of two parts: coder and decoder. The task of
```
Comment (montekki) [inline content/posts/detecting-web-attacks-rnn.md]: This is a re-word of the above, you can delete this

File: content/posts/detecting-web-attacks-rnn.md [hunk 3]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+to the original form. If the certain input differs greatly from other data, the decoder
+can't reconstruct a representation correctly.
+
+That means that the model has never seen the data like this. So this input is considered
+as an anomaly.
+
 ![autoencoder](images/detecting-web-attacks-rnn-02.png)
 
-### The result
+*image taken from [What to do when data is missing, Part II](http://curiousily.com/data-science/2017/02/02/what-to-do-when-data-is-missing-part-2.html)*
```
Comment (montekki) [inline content/posts/detecting-web-attacks-rnn.md]: this shouldn't be here, probably a mistake during a merge.

File: content/posts/detecting-web-attacks-rnn.md [hunk 4]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
 
-Finally, we ended up with a seq2seq autoencoder model that proved to be able to find
-anomalies in HTTP requests. Pre-processed request data is fed both to inputs
-and the outputs of the model, thus it learns to re-create the sequences it sees.
+Our solution is made up of several parts: model definition, training part, prediction
+part for setting specific constants and validation.
+
+We avoid typical code in the [solution](https://github.com/PositiveTechnologies/seq2seq-web-attack-detection) and comment on a few aspects only.
```
Comment (montekki) [inline content/posts/detecting-web-attacks-rnn.md]: avoid?

File: content/posts/detecting-web-attacks-rnn.md [hunk 5]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+use the probabilities to mark the malicious parts of a given request.
+
+[attention](images/malicious.png)
+
+At the testing stage with our samples we've got very good results: precision and
+recall are about 0.99. And the ROC curve tends to be 1. Looks amazing!
+
+[ROC](images/roc.png)
+
+Let's just break down what we know already...
```
Comment (montekki) [inline content/posts/detecting-web-attacks-rnn.md]: Please don't submit big pieces of text that will bring a second writer style to the text. Imagine reading a mix of Gogol and Tolstoy

File: content/posts/detecting-web-attacks-rnn.md (hunk 1 above)
Comment (AlexandraMur) [reply content/posts/detecting-web-attacks-rnn.md]: I'm sure the grammar is correct. And I'd like to focus on that we consider just a single request.

File: content/posts/detecting-web-attacks-rnn.md (hunk 2 above)
Comment (AlexandraMur) [reply content/posts/detecting-web-attacks-rnn.md]: I think it's ok. Previous part was too formal and here we simply explain the core of autoencoder.

File: content/posts/detecting-web-attacks-rnn.md (hunk 5 above)
Comment (AlexandraMur) [reply content/posts/detecting-web-attacks-rnn.md]: I'm intentionally trying to bring some informal part to make the text more simple.

File: content/posts/detecting-web-attacks-rnn.md [hunk 6]
```diff
@@ -27,21 +27,33 @@ used to back in a day. But does it?
...
-embedded into a reverse proxy of some type. The most prominent example is mod_security, 
+Most WAFs nowadays attempt to detect attacks in a similar fashion: some rule-based engine
+embedded into a reverse proxy of some type. The most prominent example is mod_security,
 a WAF module for Apache web server, which was created in 2002. Rule-based detection
 has some disadvantages, for instance, they fail to detect novel attacks aka 0days while these
 same attacks can easily be detected by a human expert. If you think about it, that's
 not very surprising, human brain works very differently from a set of regular
 expressions.
 
+From perspectives of WAFs, the attack types can be divided on Time series-based
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 7]
```diff
@@ -27,21 +27,33 @@ used to back in a day. But does it?
...
+and a single HTTP request/response-based. We consider only single requests. In this
+case we able to detect attacks like this:
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 8]
```diff
@@ -96,6 +108,26 @@ it is.
 Essentially, our goal is to make our attack detection AI work in some fashion
 that resembles this human reasoning.
 
+A tricky moment there is that some traffic seem to be malicious can be typical for
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 9]
```diff
@@ -96,6 +108,26 @@ it is.
...
+a specific web server.
+
+So for instance, let's take a look at this request:
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 10]
```diff
@@ -96,6 +108,26 @@ it is.
...
+[malicious](images/b_request1.png)
+
+Is it anomaly?
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 11]
```diff
@@ -96,6 +108,26 @@ it is.
...
+Actually this request of the bag tracker and such a request is typical for this service.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 12]
```diff
@@ -96,6 +108,26 @@ it is.
...
+It means that the request is benign.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 13]
```diff
@@ -96,6 +108,26 @@ it is.
...
+
+And let's look at another case:
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 14]
```diff
@@ -96,6 +108,26 @@ it is.
...
+[Joomla1](images/m_request1.png)
+
+At first sight, it seemed to be ok. But parameter isn't typical for this request,
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 15]
```diff
@@ -96,6 +108,26 @@ it is.
...
+albeit definitely felt like benign, it's not even a quote sign.
+
+And this "benign" parameter attacks Joomla < 3.6.4 (Privilege Elevation).
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 16]
```diff
@@ -131,6 +163,10 @@ For instance, natural language classification RNNs use word embeddings, however
 it is not clear what words there are in a non-natural language like HTTP. That's why
 we decided to use character embeddings in our model.
 
+We didn't use complicated embedding just decided matching 'chars' to numbers. All possible
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 17]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
 is given an anomalous sample it is likely to re-create it with a high degree
 of error.
 
+So, basically, autoencoder model consists of two parts: coder and decoder. The task of
+the encoder is to compress the input data into a special representation. The task
+of the decoder part is to learn how to decode correctly from a special representation
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 18]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+to the original form. If the certain input differs greatly from other data, the decoder
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 19]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+can't reconstruct a representation correctly.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 20]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+That means that the model has never seen the data like this. So this input is considered
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 21]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
 ![autoencoder](images/detecting-web-attacks-rnn-02.png)
 
-### The result
+*image taken from [What to do when data is missing, Part II](http://curiousily.com/data-science/2017/02/02/what-to-do-when-data-is-missing-part-2.html)*
 
-Finally, we ended up with a seq2seq autoencoder model that proved to be able to find
-anomalies in HTTP requests. Pre-processed request data is fed both to inputs
-and the outputs of the model, thus it learns to re-create the sequences it sees.
+Our solution is made up of several parts: model definition, training part, prediction
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 22]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+part for setting specific constants and validation.
+
+We avoid explanation of typical code in the [solution](https://github.com/PositiveTechnologies/seq2seq-web-attack-detection) and comment on a few aspects only.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 23]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+For working with raw data we made several useful utils. The code is really simple
+and obvious, so, we hid its realization in the baseline and took it to the data folder.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 24]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+In the model definition part we define `Seq2Seq` class where in `__init__` method
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 25]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+num_layers - the number of lstm blocks
+checkpoints - path to checkpoint directory
+std_factor - the number of stds that is used for defining a model threshold
+dropout - the probability that each element is kept
+vocab - the Vocabulary object
+```
 
-We built our model into our WAF and trained against our test vulnerable application
-of a bank.
+Then we define autoencoder's layers. For the encoder's part this way:
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 26]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+  _, enc_state = tf.nn.dynamic_rnn(
+      multilstm,
+      enc_embed_input,
+      sequence_length=self.lengths,
+      swap_memory=True,
+      dtype=tf.float32)
+```
+
+The decoder's part like that:
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 27]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, enc_state)
+
+dec_outputs = tf.contrib.seq2seq.dynamic_decode(
+    decoder,
+    output_time_major=False,
+    impute_finished=True,
+    maximum_iterations=self.max_seq_len, swap_memory=True)
+```
+
+In short, the training process is quite typical for training models. We just fit in
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 28]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+our model on each step and run the tensorflow session.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 29]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+In case of anomaly detection problem we reconstruct input data, so targets and
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 30]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+feed_dict = {
+  model.inputs: X,
+  model.targets: X,
+  model.lengths: L,
+  model.dropout: self.dropout,
+  model.batch_size: self.batch_size,
+  model.max_seq_len: seq_len}
+```
+
+The best model version we save as a checkpoint. Then at the stage of prediction,
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 31]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+we load it. In our proof of concept, we had not only Jupiter notebook but a running
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 32]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+web application where we can load model and manually exploit some vulnerabilities
+to test our solution better.
+
+Being inspired by the attention mechanism we tried to apply it to autoencoder but
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 33]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+noticed that probabilities output from the last layer works better. We began to
+use the probabilities to mark the malicious parts of a given request.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 34]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+[attention](images/malicious.png)
+
+At the testing stage with our samples we've got very good results: precision and
+recall are about 0.99. And the ROC curve tends to be 1. Looks amazing!
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 35]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+[ROC](images/roc.png)
+
+Let's just break down what we know already...
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 36]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+it is much different from being request.
+
+Ok, well.. what if we face something malicious directly in the parameter value or
+maybe in the parameter name... Here is lstm layers help us. Actually, we learn not
+only what is typical for requests but also we learn how each character is typically
+relative to one another. And that's the reason why we don't need complicated embedding.
+And we learn some relation on entire sequence length for each position
+relative to each other for traffic which is typical for your specific application.
+
+The complete proof of concept you can see there: [github.com](https://github.com/PositiveTechnologies/seq2seq-web-attack-detection)
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 37]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+
+### The results
+
+Finally, we ended up with a seq2seq autoencoder model that proved to be able to find
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md [hunk 38]
```diff
@@ -172,16 +208,131 @@ in other words, approximate an identity function. If the trained autoencoder
...
+anomalies in HTTP requests. We built our model into our WAF and trained against
+our test vulnerable application of a bank.
```
[the remark this answers is not in the document]
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: fixed

File: content/posts/detecting-web-attacks-rnn.md (hunk 3 above)
Comment (amurzina) [reply content/posts/detecting-web-attacks-rnn.md]: Fix it yourself, please.

**Pull request merged**